quadcontrol/config.c

```diff
@@ -277,3 +297,233 @@ int config_scenarioRead(const char *path, flight_mode_t **scenario, size_t *sz)
 
 	return 0;
 }
+
+
+static int config_pidConverter(const hmap_t *h)
+{
+	int err = 0;
+	pid_ctx_t *pid;
+	int id = res[cfg_pidID].invCnt;
+
+	if (config_handleSpace(cfg_pidID, sizeof(pid_ctx_t)) != 0) {
+		return -1;
+	}
+
+	pid = (pid_ctx_t *)res[cfg_pidID].data + id;
+
+	err |= config_parseFloat(h, "P", &pid->kp);
+	err |= config_parseFloat(h, "I", &pid->ki);
+	err |= config_parseFloat(h, "D", &pid->kd);
+	err |= config_parseFloat(h, "MIN", &pid->min);
+	err |= config_parseFloat(h, "MAX", &pid->max);
+	err |= config_parseFloat(h, "IMAX", &pid->maxInteg);
+	err |= config_parseFloat(h, "IMIN", &pid->minInteg);
+
+	if (err != 0) {
+		return -1;
+	}
+
+	res[cfg_pidID].invCnt++;
+
+	return 0;
+}
+
+
+int config_pidRead(const char *path, pid_ctx_t **pids, int *sz)
+{
+	int err;
+	parser_t *p;
+	static const unsigned int initSz = 4;
+
+	if (path == NULL || pids == NULL || sz == NULL) {
+		fprintf(stderr, "config: invalid arguments\n");
+		return -1;
+	}
+
+	p = parser_alloc(1, 7);
+	if (p == NULL) {
+		return -1;
+	}
+
+	if (parser_headerAdd(p, "PID", config_pidConverter) != 0) {
+		parser_free(p);
+		return -1;
+	}
+
+	res[cfg_pidID].data = malloc(sizeof(pid_ctx_t) * initSz);
+	if (res[cfg_pidID].data == NULL) {
+		parser_free(p);
+		return -1;
+	}
+
+	res[cfg_pidID].sz = initSz;
+	res[cfg_pidID].invCnt = 0;
+
+	err = parser_execute(p, path, PARSER_IGN_UNKNOWN_HEADERS);
+	parser_free(p);
+	if (err != 0) {
+		free(res[cfg_pidID].data);
+		return -1;
+	}
+
+	if (config_adjustSpace(cfg_pidID, sizeof(pid_ctx_t)) != 0) {
+		fprintf(stderr, "config: realloc error\n");
+		free(res[cfg_pidID].data);
+		return -1;
+	}
+
+	*pids = res[cfg_pidID].data;
+	*sz = res[cfg_pidID].sz;
+
+	return 0;
+}
+
+
+static int config_throttleConverter(const hmap_t *h)
+{
+	int err = 0;
+	quad_throttle_t *throttle;
+	int id = res[cfg_throttleID].invCnt;
+
+	if (config_handleSpace(cfg_throttleID, sizeof(quad_throttle_t)) != 0) {
+		return -1;
+	}
+
+	throttle = (quad_throttle_t *)res[cfg_throttleID].data + id;
+
+	err |= config_parseFloat(h, "MAX", &throttle->max);
+	err |= config_parseFloat(h, "MIN", &throttle->min);
+
+	if (err != 0) {
+		return -1;
+	}
+
+	res[cfg_throttleID].invCnt++;
+
+	return 0;
+}
+
+
+int config_throttleRead(const char *path, quad_throttle_t **throttle, int *sz)
+{
+	int err;
+	parser_t *p;
+	static const unsigned int initSz = 1;
+
+	if (path == NULL || throttle == NULL || sz == NULL) {
+		fprintf(stderr, "config: invalid arguments\n");
+		return -1;
+	}
+
```

Comment:
You can apply this for other functions:

```suggestion
*throttle = NULL;
*sz = 0;
```